BBC: Blead Breaks IO::Async

Passes on 5.37.8.

On blead:

```
$ bleadperl -Mblib t/41routine.t 
ok 1 - '$routine for fork model' isa 'IO::Async::Routine'
ok 2 - $routine has refcount 1 initially for fork model
ok 3 - $routine has refcount 2 after $loop->add for fork model
ok 4 - $routine->model for fork model
ok 5 - Result for fork model
ok 6 - $routine has refcount 2 before $loop->remove
ok 7 - $routine has refcount 1 before EOF
ok 8 - on_return for fork model
ok 9 - Result from ->result_future of return
ok 10 - on_die for fork model
ok 11 - Failure from ->result_future of die
ok 12 - Recv on closed channel for fork model
ok 13 - Routine is ready for SIGINT
ok 14 - Routine caught SIGINT
ok 15 - '$routine for thread model' isa 'IO::Async::Routine'
ok 16 - $routine has refcount 1 initially for thread model
Segmentation fault
```

Appears to be a stack blowup:

```
$ bleadperl --gdb -Mblib t/41routine.t 
...
(gdb) run
Starting program: /home/leo/perl5/perlbrew/perls/bleadperl/bin/perl -Mblib t/41routine.t
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
ok 1 - '$routine for fork model' isa 'IO::Async::Routine'
ok 2 - $routine has refcount 1 initially for fork model
[Detaching after fork from child process 2491048]
ok 3 - $routine has refcount 2 after $loop->add for fork model
ok 4 - $routine->model for fork model
ok 5 - Result for fork model
ok 6 - $routine has refcount 2 before $loop->remove
ok 7 - $routine has refcount 1 before EOF
[Detaching after fork from child process 2491049]
ok 8 - on_return for fork model
ok 9 - Result from ->result_future of return
[Detaching after fork from child process 2491050]
ok 10 - on_die for fork model
ok 11 - Failure from ->result_future of die
[Detaching after fork from child process 2491051]
ok 12 - Recv on closed channel for fork model
[Detaching after fork from child process 2491052]
ok 13 - Routine is ready for SIGINT
ok 14 - Routine caught SIGINT
ok 15 - '$routine for thread model' isa 'IO::Async::Routine'
ok 16 - $routine has refcount 1 initially for thread model

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7d39895 in __GI___libc_malloc (bytes=40) at ./malloc/malloc.c:3315
3315    ./malloc/malloc.c: No such file or directory.
(gdb) bt
#0  0x00007ffff7d39895 in __GI___libc_malloc (bytes=40) at ./malloc/malloc.c:3315
#1  0x00005555557a291a in Perl_safesysmalloc (size=40) at util.c:161
#2  0x0000555555833784 in Perl_sv_grow_fresh (my_perl=0x555556cbdb20, sv=0x5555573f5478, newlen=16) at sv.c:1421
#3  0x0000555555858217 in Perl_sv_setpvn_fresh (my_perl=0x555556cbdb20, sv=0x5555573f5478, ptr=0x555555bc99f4 "", len=0) at sv.c:5029
#4  0x000055555587564e in Perl_newSVpvn_flags (my_perl=0x555556cbdb20, s=0x555555bc99f4 "", len=0, flags=524288) at sv.c:9551
#5  0x00005555557a53c8 in S_mess_alloc (my_perl=0x555556cbdb20) at util.c:1383
#6  0x00005555557a67a5 in Perl_vmess (my_perl=0x555556cbdb20, pat=0x555555bff11a "%s", args=0x7fffff7ff280) at util.c:1673
#7  0x00005555557a7d5b in Perl_vcroak (my_perl=0x555556cbdb20, pat=0x555555bff11a "%s", args=0x7fffff7ff280) at util.c:1892
#8  0x00005555557a80cf in Perl_croak_nocontext (pat=0x555555bff11a "%s") at util.c:1934
#9  0x00005555558d52dc in Perl_croak_memory_wrap () at perlstatic.h:24
#10 0x00005555558d6fb2 in Perl_tmps_grow_p (my_perl=0x555556cbdb20, ix=-6076574518398426621) at scope.c:218
#11 0x00005555558756b7 in Perl_newSVpvn_flags (my_perl=0x555556cbdb20, s=0x555555bc99f4 "", len=0, flags=524288) at sv.c:9564
#12 0x00005555557a53c8 in S_mess_alloc (my_perl=0x555556cbdb20) at util.c:1383
#13 0x00005555557a67a5 in Perl_vmess (my_perl=0x555556cbdb20, pat=0x555555bff11a "%s", args=0x7fffff7ff4d0) at util.c:1673
#14 0x00005555557a7d5b in Perl_vcroak (my_perl=0x555556cbdb20, pat=0x555555bff11a "%s", args=0x7fffff7ff4d0) at util.c:1892
#15 0x00005555557a80cf in Perl_croak_nocontext (pat=0x555555bff11a "%s") at util.c:1934
#16 0x00005555558d52dc in Perl_croak_memory_wrap () at perlstatic.h:24
#17 0x00005555558d6fb2 in Perl_tmps_grow_p (my_perl=0x555556cbdb20, ix=-6076574518398426622) at scope.c:218
#18 0x00005555558756b7 in Perl_newSVpvn_flags (my_perl=0x555556cbdb20, s=0x555555bc99f4 "", len=0, flags=524288) at sv.c:9564
#19 0x00005555557a53c8 in S_mess_alloc (my_perl=0x555556cbdb20) at util.c:1383
#20 0x00005555557a67a5 in Perl_vmess (my_perl=0x555556cbdb20, pat=0x555555bff11a "%s", args=0x7fffff7ff720) at util.c:1673
#21 0x00005555557a7d5b in Perl_vcroak (my_perl=0x555556cbdb20, pat=0x555555bff11a "%s", args=0x7fffff7ff720) at util.c:1892
#22 0x00005555557a80cf in Perl_croak_nocontext (pat=0x555555bff11a "%s") at util.c:1934
#23 0x00005555558d52dc in Perl_croak_memory_wrap () at perlstatic.h:24
#24 0x00005555558d6fb2 in Perl_tmps_grow_p (my_perl=0x555556cbdb20, ix=-6076574518398426623) at scope.c:218
#25 0x00005555558756b7 in Perl_newSVpvn_flags (my_perl=0x555556cbdb20, s=0x555555bc99f4 "", len=0, flags=524288) at sv.c:9564
...
(continues for a long time)
```

In particular it appears to be stuck in infinite recursion trying to croak.

I admit I am now kinda stuck. With the latest commit on my branch (see footnote) I have added a "WIP" failing test, which creates/uses/destroys an object instance entirely within a short-lived thread and exits.

```perl
threads->create(sub {
    my $obj = Test1->new(x => 20);
    is($obj->x, 20, '$obj->x created inside thread');
})->join;
```

That test itself passes, along with thread join, but it leaves behind some sort of junk that later causes global destruction to fail:

```
$ gdb --args ./perl -Ilib -MDevel::MAT::Dumper t/class/threads.t 
...
ok 2 - $obj->x created inside thread
[Thread 0x7ffff77ff6c0 (LWP 2721754) exited]
1..1
panic: free from wrong pool, 5555560c9980!=555555c7b2a0 during global destruction.
```

This panic doesn't happen without that test.

On further inspection of the stack (which is why my WIP commit switches it to `raise(SIGABRT)`) I see:

```
(gdb) bt
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0)
    at ./nptl/pthread_kill.c:44
#1  0x00007ffff7d2bd2f in __pthread_kill_internal (signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:78
#2  0x00007ffff7cdcef2 in __GI_raise (sig=6) at ../sysdeps/posix/raise.c:26
#3  0x00005555557a2e74 in Perl_safesysfree (where=0x7ffff0001c28) at util.c:380
#4  0x00005555558668d2 in Perl_sv_clear (my_perl=0x555555c7b2a0, orig_sv=0x5555560cdc88) at sv.c:6775
#5  0x0000555555869a0e in Perl_sv_free2 (my_perl=0x555555c7b2a0, sv=0x5555560cdc88, rc=1) at sv.c:7228
....
```

So it was trying to free the SV at 0x5555560cdc88. I can now use `pmat` to work out what that SV was (which is why I added Devel::MAT::Dumper to the test):

```
$ pmat t/perl.pmat 
Loading file t/perl.pmat...
...

pmat> show 0x5555560cdc88
ARRAY(0,!REAL) at 0x5555560cdc88 with refcount 1
  size 72 bytes
  0 elements (use 'elems' command to show)

pmat> identify 0x5555560cdc88
ARRAY(0,!REAL) at 0x5555560cdc88 is:
└─the @_ av at depth 1 of CODE(PP) at 0x555555cc0838, which is:
  └─the symbol '&Test1::x'
```

This shows that the SV it's upset about freeing, is the `@_` array of the `Test1::x` method.

At this point I'm kinda stuck unsure what to look at further or why it's even upset in the first place.

Anyone any thoughts?

-----

My branch: https://github.com/leonerd/perl5/tree/feature-class-with-threads

--- a/MANIFEST
+++ b/MANIFEST
@@ -5648,6 +5648,7 @@ t/class/field.t			See if class field declarations work
 t/class/inherit.t		See if class inheritance works
 t/class/method.t		See if class method declarations work
 t/class/phasers.t		See if class phaser blocks work
+t/class/threads.t		See if classes work across multiple threads
 t/cmd/elsif.t			See if else-if works
 t/cmd/for.t			See if for loops work
 t/cmd/mod.t			See if statement modifiers work

--- a/class.c
+++ b/class.c
@@ -129,7 +129,7 @@ XS(injected_constructor)
 {
     dXSARGS;
 
-    HV *stash = (HV *)XSANY.any_ptr;
+    HV *stash = (HV *)XSANY.any_sv;
     assert(HvSTASH_IS_CLASS(stash));
 
     struct xpvhv_aux *aux = HvAUX(stash);
@@ -353,7 +353,8 @@ Perl_class_setup_stash(pTHX_ HV *stash)
         SAVEFREESV(newname);
 
         CV *newcv = newXS_flags(SvPV_nolen(newname), injected_constructor, __FILE__, NULL, nameflags);
-        CvXSUBANY(newcv).any_ptr = stash;
+        CvXSUBANY(newcv).any_sv = (SV *)stash;
+        CvREFCOUNTED_ANYSV_on(newcv);
     }
 
     /* TODO:

--- a/embed.fnc
+++ b/embed.fnc
@@ -5656,6 +5656,9 @@ S	|SV **	|sv_dup_inc_multiple					\
 				|NN SV **dest				\
 				|SSize_t items				\
 				|NN CLONE_PARAMS * const param
+S	|void	|sv_dup_hvaux	|NN const SV * const ssv		\
+				|NN SV *dsv				\
+				|NN CLONE_PARAMS * const param
 SR	|SV *	|sv_dup_common	|NN const SV * const ssv		\
 				|NN CLONE_PARAMS * const param
 S	|void	|unreferenced_to_tmp_stack				\

--- a/embed.h
+++ b/embed.h
@@ -2055,6 +2055,7 @@
 #     endif /* defined(PERL_DEBUG_READONLY_COW) */
 #     if defined(USE_ITHREADS)
 #       define sv_dup_common(a,b)               S_sv_dup_common(aTHX_ a,b)
+#       define sv_dup_hvaux(a,b,c)              S_sv_dup_hvaux(aTHX_ a,b,c)
 #       define sv_dup_inc_multiple(a,b,c,d)     S_sv_dup_inc_multiple(aTHX_ a,b,c,d)
 #       define unreferenced_to_tmp_stack(a)     S_unreferenced_to_tmp_stack(aTHX_ a)
 #     endif /* defined(USE_ITHREADS) */

--- a/pad.c
+++ b/pad.c
@@ -2864,6 +2864,16 @@ Perl_padname_dup(pTHX_ PADNAME *src, CLONE_PARAMS *param)
     PadnameTYPE   (dst) = (HV *)sv_dup_inc((SV *)PadnameTYPE(src), param);
     PadnameOURSTASH(dst) = (HV *)sv_dup_inc((SV *)PadnameOURSTASH(src),
                                             param);
+    if(PadnameIsFIELD(src)) {
+        struct padname_fieldinfo *sinfo = PadnameFIELDINFO(src);
+        struct padname_fieldinfo *dinfo;
+        Newxz(dinfo, 1, struct padname_fieldinfo);
+
+        dinfo->fieldix    = sinfo->fieldix;
+        dinfo->fieldstash = hv_dup_inc(sinfo->fieldstash, param);
+
+        PadnameFIELDINFO(dst) = dinfo;
+    }
     dst->xpadn_low  = src->xpadn_low;
     dst->xpadn_high = src->xpadn_high;
     dst->xpadn_gen  = src->xpadn_gen;

--- a/proto.h
+++ b/proto.h
@@ -9195,6 +9195,11 @@ S_sv_dup_common(pTHX_ const SV * const ssv, CLONE_PARAMS * const param)
 #   define PERL_ARGS_ASSERT_SV_DUP_COMMON       \
         assert(ssv); assert(param)
 
+STATIC void
+S_sv_dup_hvaux(pTHX_ const SV * const ssv, SV *dsv, CLONE_PARAMS * const param);
+#   define PERL_ARGS_ASSERT_SV_DUP_HVAUX        \
+        assert(ssv); assert(dsv); assert(param)
+
 STATIC SV **
 S_sv_dup_inc_multiple(pTHX_ SV * const *source, SV **dest, SSize_t items, CLONE_PARAMS * const param);
 #   define PERL_ARGS_ASSERT_SV_DUP_INC_MULTIPLE \

--- a/sv.c
+++ b/sv.c
@@ -14423,6 +14423,72 @@ S_sv_dup_inc_multiple(pTHX_ SV *const *source, SV **dest,
     return dest;
 }
 
+/* duplicate the HvAUX of an HV */
+static void
+S_sv_dup_hvaux(pTHX_ const SV *const ssv, SV *dsv, CLONE_PARAMS *const param)
+{
+    PERL_ARGS_ASSERT_SV_DUP_HVAUX;
+
+    const struct xpvhv_aux * const saux = HvAUX(ssv);
+    struct xpvhv_aux * const daux = HvAUX(dsv);
+    /* This flag isn't copied.  */
+    SvOOK_on(dsv);
+
+    if (saux->xhv_name_count) {
+        HEK ** const sname = saux->xhv_name_u.xhvnameu_names;
+        const I32 count = saux->xhv_name_count < 0
+            ? -saux->xhv_name_count
+            :  saux->xhv_name_count;
+        HEK **shekp = sname + count;
+        HEK **dhekp;
+        Newx(daux->xhv_name_u.xhvnameu_names, count, HEK *);
+        dhekp = daux->xhv_name_u.xhvnameu_names + count;
+        while (shekp-- > sname) {
+            dhekp--;
+            *dhekp = hek_dup(*shekp, param);
+        }
+    }
+    else {
+        daux->xhv_name_u.xhvnameu_name = hek_dup(saux->xhv_name_u.xhvnameu_name, param);
+    }
+    daux->xhv_name_count = saux->xhv_name_count;
+
+    daux->xhv_aux_flags = saux->xhv_aux_flags;
+#ifdef PERL_HASH_RANDOMIZE_KEYS
+    daux->xhv_rand = saux->xhv_rand;
+    daux->xhv_last_rand = saux->xhv_last_rand;
+#endif
+    daux->xhv_riter = saux->xhv_riter;
+    daux->xhv_eiter = saux->xhv_eiter ? he_dup(saux->xhv_eiter, FALSE, param) : 0;
+    /* backref array needs refcnt=2; see sv_add_backref */
+    daux->xhv_backreferences =
+        (param->flags & CLONEf_JOIN_IN)
+            /* when joining, we let the individual GVs and
+             * CVs add themselves to backref as
+             * needed. This avoids pulling in stuff
+             * that isn't required, and simplifies the
+             * case where stashes aren't cloned back
+             * if they already exist in the parent
+             * thread */
+        ? NULL
+        : saux->xhv_backreferences
+            ? (SvTYPE(saux->xhv_backreferences) == SVt_PVAV)
+                ? MUTABLE_AV(SvREFCNT_inc(
+                      sv_dup_inc((const SV *)
+                        saux->xhv_backreferences, param)))
+                : MUTABLE_AV(sv_dup((const SV *)
+                        saux->xhv_backreferences, param))
+            : 0;
+
+    daux->xhv_mro_meta = saux->xhv_mro_meta
+        ? mro_meta_dup(saux->xhv_mro_meta, param)
+        : 0;
+
+    /* Record stashes for possible cloning in Perl_clone(). */
+    if (HvNAME(ssv))
+        av_push(param->stashes, dsv);
+}
+
 /* duplicate an SV of any type (including AV, HV etc) */
 
 static SV *
@@ -14731,70 +14797,8 @@ S_sv_dup_common(pTHX_ const SV *const ssv, CLONE_PARAMS *const param)
                             ? he_dup(source, FALSE, param) : 0;
                         ++i;
                     }
-                    if (HvHasAUX(ssv)) {
-                        const struct xpvhv_aux * const saux = HvAUX(ssv);
-                        struct xpvhv_aux * const daux = HvAUX(dsv);
-                        /* This flag isn't copied.  */
-                        SvOOK_on(dsv);
-
-                        if (saux->xhv_name_count) {
-                            HEK ** const sname = saux->xhv_name_u.xhvnameu_names;
-                            const I32 count
-                             = saux->xhv_name_count < 0
-                                ? -saux->xhv_name_count
-                                :  saux->xhv_name_count;
-                            HEK **shekp = sname + count;
-                            HEK **dhekp;
-                            Newx(daux->xhv_name_u.xhvnameu_names, count, HEK *);
-                            dhekp = daux->xhv_name_u.xhvnameu_names + count;
-                            while (shekp-- > sname) {
-                                dhekp--;
-                                *dhekp = hek_dup(*shekp, param);
-                            }
-                        }
-                        else {
-                            daux->xhv_name_u.xhvnameu_name
-                                = hek_dup(saux->xhv_name_u.xhvnameu_name,
-                                          param);
-                        }
-                        daux->xhv_name_count = saux->xhv_name_count;
-
-                        daux->xhv_aux_flags = saux->xhv_aux_flags;
-#ifdef PERL_HASH_RANDOMIZE_KEYS
-                        daux->xhv_rand = saux->xhv_rand;
-                        daux->xhv_last_rand = saux->xhv_last_rand;
-#endif
-                        daux->xhv_riter = saux->xhv_riter;
-                        daux->xhv_eiter = saux->xhv_eiter
-                            ? he_dup(saux->xhv_eiter, FALSE, param) : 0;
-                        /* backref array needs refcnt=2; see sv_add_backref */
-                        daux->xhv_backreferences =
-                            (param->flags & CLONEf_JOIN_IN)
-                                /* when joining, we let the individual GVs and
-                                 * CVs add themselves to backref as
-                                 * needed. This avoids pulling in stuff
-                                 * that isn't required, and simplifies the
-                                 * case where stashes aren't cloned back
-                                 * if they already exist in the parent
-                                 * thread */
-                            ? NULL
-                            : saux->xhv_backreferences
-                                ? (SvTYPE(saux->xhv_backreferences) == SVt_PVAV)
-                                    ? MUTABLE_AV(SvREFCNT_inc(
-                                          sv_dup_inc((const SV *)
-                                            saux->xhv_backreferences, param)))
-                                    : MUTABLE_AV(sv_dup((const SV *)
-                                            saux->xhv_backreferences, param))
-                                : 0;
-
-                        daux->xhv_mro_meta = saux->xhv_mro_meta
-                            ? mro_meta_dup(saux->xhv_mro_meta, param)
-                            : 0;
-
-                        /* Record stashes for possible cloning in Perl_clone(). */
-                        if (HvNAME(ssv))
-                            av_push(param->stashes, dsv);
-                    }
+                    if (HvHasAUX(ssv))
+                        sv_dup_hvaux(ssv, dsv, param);
                 }
                 else
                     HvARRAY(MUTABLE_HV(dsv)) = NULL;

--- a/sv.c
+++ b/sv.c
@@ -14432,7 +14432,7 @@ S_sv_dup_hvaux(pTHX_ const SV *const ssv, SV *dsv, CLONE_PARAMS *const param)
     const struct xpvhv_aux * const saux = HvAUX(ssv);
     struct xpvhv_aux * const daux = HvAUX(dsv);
     /* This flag isn't copied.  */
-    SvOOK_on(dsv);
+    SvFLAGS(dsv) |= SVphv_HasAUX;
 
     if (saux->xhv_name_count) {
         HEK ** const sname = saux->xhv_name_u.xhvnameu_names;

--- a/sv.c
+++ b/sv.c
@@ -14487,6 +14487,15 @@ S_sv_dup_hvaux(pTHX_ const SV *const ssv, SV *dsv, CLONE_PARAMS *const param)
     /* Record stashes for possible cloning in Perl_clone(). */
     if (HvNAME(ssv))
         av_push(param->stashes, dsv);
+
+    if (HvSTASH_IS_CLASS(ssv)) {
+        daux->xhv_class_superclass    = hv_dup_inc(saux->xhv_class_superclass,    param);
+        daux->xhv_class_initfields_cv = cv_dup_inc(saux->xhv_class_initfields_cv, param);
+        daux->xhv_class_adjust_blocks = av_dup_inc(saux->xhv_class_adjust_blocks, param);
+        daux->xhv_class_fields        = padnamelist_dup_inc(saux->xhv_class_fields, param);
+        daux->xhv_class_next_fieldix  = saux->xhv_class_next_fieldix;
+        daux->xhv_class_param_map     = hv_dup_inc(saux->xhv_class_param_map,     param);
+    }
 }
 
 /* duplicate an SV of any type (including AV, HV etc) */
@@ -14623,6 +14632,7 @@ S_sv_dup_common(pTHX_ const SV *const ssv, CLONE_PARAMS *const param)
                     goto have_body;
                 }
                 /* FALLTHROUGH */
+            case SVt_PVOBJ:
             case SVt_PVGV:
             case SVt_PVIO:
             case SVt_PVFM:
@@ -14661,7 +14671,7 @@ S_sv_dup_common(pTHX_ const SV *const ssv, CLONE_PARAMS *const param)
                  sv_type_details->body_size + sv_type_details->offset, char);
 #endif
 
-            if (sv_type != SVt_PVAV && sv_type != SVt_PVHV
+            if (sv_type != SVt_PVAV && sv_type != SVt_PVHV && sv_type != SVt_PVOBJ
                 && !isGV_with_GP(dsv)
                 && !isREGEXP(dsv)
                 && !(sv_type == SVt_PVIO && !(IoFLAGS(dsv) & IOf_FAKE_DIRP)))
@@ -14854,6 +14864,16 @@ S_sv_dup_common(pTHX_ const SV *const ssv, CLONE_PARAMS *const param)
                     CvWEAKOUTSIDE(ssv)
                     ? cv_dup(    CvOUTSIDE(dsv), param)
                     : cv_dup_inc(CvOUTSIDE(dsv), param);
+                break;
+            case SVt_PVOBJ:
+                {
+                    Size_t fieldcount = ObjectMAXFIELD(ssv) + 1;
+
+                    Newx(ObjectFIELDS(dsv), fieldcount, SV *);
+                    ObjectMAXFIELD(dsv) = fieldcount - 1;
+
+                    sv_dup_inc_multiple(ObjectFIELDS(ssv), ObjectFIELDS(dsv), fieldcount, param);
+                }
                 break;
             }
         }

--- a/t/class/threads.t
+++ b/t/class/threads.t
@@ -1,0 +1,47 @@
+#!./perl
+
+BEGIN {
+    chdir 't' if -d 't';
+    require './test.pl';
+    set_up_inc('../lib');
+    require Config;
+
+    skip_all_without_config('useithreads');
+    skip_all_if_miniperl("no dynamic loading on miniperl, no threads");
+}
+
+use v5.36;
+use feature 'class';
+no warnings 'experimental::class';
+
+use threads;
+
+class Test1 {
+    field $x :param;
+    method x { return $x }
+}
+
+{
+    my $ret = threads->create(sub {
+        pass("Created dummy thread");
+        return 1;
+    })->join;
+    next_test(); # account for pass() inside thread
+    is($ret, 1, "Returned from dummy thread");
+}
+
+{
+    my $obj = Test1->new(x => 10);
+    threads->create(sub {
+        is($obj->x, 10, '$obj->x inside thread created before');
+    })->join;
+    next_test(); # account for is() inside thread
+}
+
+threads->create(sub {
+    my $obj = Test1->new(x => 20);
+    is($obj->x, 20, '$obj->x created inside thread');
+})->join;
+next_test(); # account for is() inside thread
+
+done_testing;

